lib/ui/painting/image_descriptor.cc

```diff
@@ -0,0 +1,112 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/painting/image_descriptor.h"
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/trace_event.h"
+#include "flutter/lib/ui/painting/codec.h"
+#include "flutter/lib/ui/painting/image_decoder.h"
+#include "flutter/lib/ui/painting/multi_frame_codec.h"
+#include "flutter/lib/ui/painting/single_frame_codec.h"
+#include "flutter/lib/ui/ui_dart_state.h"
+#include "third_party/tonic/dart_binding_macros.h"
+#include "third_party/tonic/logging/dart_invoke.h"
+
+namespace flutter {
+
+IMPLEMENT_WRAPPERTYPEINFO(ui, ImageDescriptor);
+
+#define FOR_EACH_BINDING(V)            \
+  V(ImageDescriptor, initRaw)          \
+  V(ImageDescriptor, instantiateCodec) \
+  V(ImageDescriptor, width)            \
+  V(ImageDescriptor, height)           \
+  V(ImageDescriptor, bytesPerPixel)
+
+FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
+
+void ImageDescriptor::RegisterNatives(tonic::DartLibraryNatives* natives) {
+  natives->Register(
+      {{"ImageDescriptor_initEncoded", ImageDescriptor::initEncoded, 3, true},
+       FOR_EACH_BINDING(DART_REGISTER_NATIVE)});
+}
+
+void ImageDescriptor::initEncoded(Dart_NativeArguments args) {
+  Dart_Handle callback_handle = Dart_GetNativeArgument(args, 2);
+  if (!Dart_IsClosure(callback_handle)) {
+    Dart_SetReturnValue(args, tonic::ToDart("Callback must be a function"));
+    return;
+  }
+
+  Dart_Handle descriptor_handle = Dart_GetNativeArgument(args, 0);
+  ImmutableBuffer* data = tonic::DartConverter<ImmutableBuffer*>::FromDart(
+      Dart_GetNativeArgument(args, 1));
+  FML_DCHECK(data);
```

Comment:
Return an error handle if this is not the case instead of dereferencing an invalid pointer later.

Reply:
Done - this should be checked on the dart side because the buffer parameter is non-nullable though.